src/odbc_result.h

```diff
@@ -24,14 +24,23 @@ class odbc_error : public Rcpp::exception {
       const std::string& sql,
       Iconv& output_encoder)
       : Rcpp::exception("", false) {
-    std::string m = std::string(e.what()) + "\n<SQL> '" + sql + "'";
+    std::string m = std::string(e.what());
+    if (sql != "" ) {
+      m += "\n<SQL> '" + sql + "'";
+    }
     // #432: [R] expects UTF-8 encoded strings but both nanodbc and sql are
     // encoded in the database encoding, which may differ from UTF-8
     message = Rf_translateChar(
         output_encoder.makeSEXP(m.c_str(), m.c_str() + m.length()));
   }
   const char* what() const NANODBC_NOEXCEPT { return message.c_str(); }
 
+  void pretty_print() const {
```

Comment:
Thinking about the names `pretty_print()` and `pretty_print_message()`... Some bits of context from R and its cli package:

* The side effects that `rethrow_database_error()` and `print_message()` are called for are conditions rather than just printed output. "Raise" or "throw" are often the verbs applied to conditions.
* In R, "messages" are a specific kind of condition (along with warnings and errors), but `pretty_print_message()` can raise either a message or warning.
* The cli functions `cli_inform()`, `cli_warn()`, and `cli_abort()` raise messages, warnings, and errors, respectively. I know this is kind of weird.

I'm wondering if we could refactor to, in Cpp,

* `raise_message()`: calls `cli::cli_inform()` directly via the `Rcpp::Environment::namespace_env("cli"); ...` pattern below
* `raise_warning()`: calls `cli::cli_warn()` directly via the same pattern
* `raise_error()`: calls `rethrow_database_error()` (which in turn calls `cli::cli_abort()`) via the current pattern

...and then writing out the `print_message()` function defined in R.

Reply:
I like it - great suggestion.  Let me know if the changes I made as part of the 'code-review' commit are not in the direction you outlined above.

Reply:
Not seeing that `code-review` commit—is it possible those changes are still local?

Reply:
Sorry about that @simonpcouch 
Hopefully managed to push to the right branch this time.

Reply:
Thanks for all of the work here! That refactor feels much better.